### PR TITLE
[skwasm] Implement platform view clipping.

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This implements platform view clipping for the Skwasm renderer.

This fixes https://github.com/flutter/flutter/issues/133466

### CODE REVIEW

File: lib/web_ui/lib/src/engine/scene_view.dart [hunk 1]
```diff
@@ -199,20 +235,14 @@ class EngineSceneView {
...
-      currentElement = sibling;
-    }
   }
 }
 
 sealed class SliceContainer {
   DomElement get container;
 
   void updateContents();
+  void dispose();
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/scene_view.dart:245]: A doc comment explaining who is responsible for calling this method would be helpful.

File: lib/web_ui/lib/src/engine/scene_view.dart [hunk 2]
```diff
@@ -301,13 +345,61 @@ final class PlatformViewContainer extends SliceContainer {
...
+    switch (clip) {
+      case PlatformViewNoClip():
+        clipPath = null;
+        return '';
+      case PlatformViewRectClip():
+        clipPath = null;
+        final double top = clip.rect.top / devicePixelRatio;
+        final double right = clip.rect.right / devicePixelRatio;
+        final double bottom = clip.rect.bottom / devicePixelRatio;
+        final double left = clip.rect.left / devicePixelRatio;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/scene_view.dart:374]: nit: it's ok to omit types in new code (or maybe even drive-by remove unnecessary types in existing code).

File: lib/web_ui/lib/src/engine/scene_view.dart [hunk 3]
```diff
@@ -42,18 +42,47 @@ class _SceneRender {
   }
 }
 
+class SVGClipRegistry {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/scene_view.dart:45]: Do we still need the SVG registry? According to MDN Safari implemented `clip-path: path(...)` in 13:

https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path

File: lib/web_ui/lib/src/engine/scene_view.dart [hunk 4]
```diff
@@ -301,13 +345,61 @@ final class PlatformViewContainer extends SliceContainer {
...
+      _clipPath = path;
+      _clipPathId = _clipRegistry.registerPath(path);
+    }
+  }
+
+  String cssStringForClip(PlatformViewClip clip, double devicePixelRatio) {
+    switch (clip) {
+      case PlatformViewNoClip():
+        clipPath = null;
+        return '';
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/scene_view.dart:368]: The function could instead return `null`, and the place that applies the clip could avoid setting the property to an empty string unnecessarily.

File: lib/web_ui/lib/src/engine/layers.dart [hunk 5]
```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
...
+class PlatformViewNoClip implements PlatformViewClip {
+  const PlatformViewNoClip();
+
+  @override
+  PlatformViewClip positioned(PlatformViewPosition positioned) {
+    return this;
+  }
+
+  @override
+  ScenePath get asPath => ui.Path() as ScenePath;
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:584]: I guess it doesn't really matter, but this seems like a more consistent and sane behavior to me than throwing. Sure, the surrounding code never calls this, but that happens to be an implementation detail of the surrounding code and not a contract of this class itself.

File: lib/web_ui/lib/src/engine/layers.dart [hunk 6]
```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
...
 }
 
+sealed class PlatformViewClip {
+  PlatformViewClip positioned(PlatformViewPosition position);
+
+  ui.Rect get boundingRect;
+
+  bool covers(ui.Rect rect);
+
+  ScenePath get asPath;
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:550]: Good point. I'll change it.

File: lib/web_ui/lib/src/engine/layers.dart [hunk 7]
```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
...
+  @override
+  bool operator ==(Object other) {
+    if (identical(this, other) || (other.runtimeType == runtimeType)) {
+      return true;
+    }
+    return false;
+  }
+
+  @override
+  int get hashCode => runtimeType.hashCode;
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:603]: I think I actually tried this, but `PlatformViewNoClip.hashCode` doesn't actually work because the language will attempt to call a static method called `hashCode` which doesn't exist. I also feel like hashing the runtime type is better than assigning an arbitrary value, because some other equally cute person might return `42` for their class which is also a singleton type, and if they end up in a very generic (i.e. `<Object>`) hashed collection together they will have a hash collision. Is this class likely to be put in a hashed collection with other random singleton classes? No. In practice it isn't even put in a hashed collection at all, but the language wants me to implement this if I implement `operator==`, so this is all theoretical and probably will never matter anyway.

File: lib/web_ui/lib/src/engine/layers.dart [hunk 8]
```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
...
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is PlatformViewRectClip && rect == other.rect;
+  }
+
+  @override
+  int get hashCode => Object.hash(runtimeType, rect);
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:647]: Well we want to avoid hash collisions in cases where they are unequal to avoid the comparison in the first place (they will be bucketed separately). I looked through our code and I found several other examples where polymorphic objects that implement `hashCode` do so by including the `runtimeType` in the hash, which seems correct and sane to me. But again, this code will probably never actually be executed and probably will just be tree-shaken out, since these are never put in a hashed collection anyway. This implementation is just to satisfy the lint.

File: lib/web_ui/lib/src/engine/layers.dart [hunk 9]
```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
...
+  bool covers(ui.Rect other) {
+    return rect.left <= other.left &&
+      rect.right >= other.right &&
+      rect.top <= other.top &&
+      rect.bottom >= other.bottom;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:640]: Hmm yeah I think I meant to do the `identical` check here.

File: lib/web_ui/lib/src/engine/layers.dart [hunk 10]
```diff
@@ -531,15 +531,211 @@ class PlatformViewStyling {
...
+
+  @override
+  ScenePath get asPath => (ui.Path() as ScenePath)..addRect(rect);
+
+  @override
+  ui.Rect get boundingRect => rect;
+
+  @override
+  bool covers(ui.Rect other) {
+    return rect.left <= other.left &&
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:632]: Actually I think we should use `rect.contains`. I'm not sure why I didn't just use it here.

File: lib/web_ui/lib/src/engine/scene_view.dart (hunk 3 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/scene_view.dart:45]: Omg seriously? Ugh, I can't believe I spent all this effort writing this class and implementing disposal methods and stuff, haha. I'll remove this and simplify.

File: lib/web_ui/lib/src/engine/scene_view.dart (hunk 4 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/scene_view.dart:368]: The container can be reused when the styling changes, so we actually *do* need to set the property to an empty string if the clipping goes away.

File: lib/web_ui/lib/src/engine/layers.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/layers.dart:632]: Ah, the reason I didn't use it is because `rect.contains(ui.Rect)` doesn't exist. `rect.contains` takes a single offset.

File: lib/web_ui/lib/src/engine/scene_view.dart [hunk 11]
```diff
@@ -390,7 +347,8 @@ final class PlatformViewContainer extends SliceContainer {
         return 'rect(${top}px ${right}px ${bottom}px ${left}px round ${tlRadiusX}px ${trRadiusX}px ${brRadiusX}px ${blRadiusX}px / ${tlRadiusY}px ${trRadiusY}px ${brRadiusY}px ${blRadiusY}px)';
       case PlatformViewPathClip():
         clipPath = clip.path;
-        return 'url(#$_clipPathId)';
+        print('$_clipPathString');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/scene_view.dart:350]: Debug print leftover?

**Review by yjbanov — APPROVED** [review]

File: lib/web_ui/lib/src/engine/scene_view.dart (hunk 11 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/scene_view.dart:350]: Whoops I'll remove it

**Comment by kevmoo** [comment]
Is there an associated issue w/ this?

**Comment by eyebrowsoffire** [comment]
> Is there an associated issue w/ this?

Yes. https://github.com/flutter/flutter/issues/133466

I'll add it to the description
